developer-workflow/extension-modules.rst

```diff
@@ -5,13 +5,579 @@
 
 * https://docs.python.org/dev/c-api/
 * https://docs.python.org/dev/extending/
 * :pep:`399`
 * https://pythonextensionpatterns.readthedocs.io/en/latest/
+
+Classifying extension modules
+=============================
+
+Extension modules can be classified into two categories:
+
+* A *built-in* extension module is a module built and shipped with
+  the Python interpreter. A built-in module is *statically* linked
+  into the interpreter, thereby lacking a :attr:`__file__` attribute.
+
+  .. seealso:: :data:`sys.builtin_module_names` --- names of built-in modules.
+
+* A *shared* (or *dynamic*) extension module is built as a shared library
+  (``.so`` or ``.dll`` file) and is *dynamically* linked into the interpreter.
+
+  In particular, the module's :attr:`__file__` attribute contains the path
+  to the ``.so`` or ``.dll`` file.
+
+.. note::
+
+   Informally, built-in extension modules can be regarded as *required*
+   while shared extension modules are *optional* in the sense that they
+   might be supplied, overridden or disabled externally.
+
+   Usually, accelerator modules are built as *shared* extension modules,
+   especially if they already have a pure Python implementation.
+
+According to :pep:`399`, *new* extension modules MUST provide a working and
+tested pure Python implementation, unless a special dispensation is given.
+Please ask the :github:`Steering Council <python/steering-council>` if such
+dispensation is needed.
+
+Adding an extension module to CPython
+=====================================
+
+Assume that the standard library contains a pure Python module :mod:`!foo`
+together with the following :func:`!foo.greet` function:
+
+.. code-block:: python
+   :caption: Lib/foo.py
+
+   def greet():
+       return "Hello World!"
+
+Instead of using the Python implementation of :func:`!foo.greet`, we want to
+use its corresponding C implementation exposed in some :mod:`!fastfoo` module
+written in C. Ideally, we want to modify ``foo.py`` as follows:
+
+.. code-block:: python
+   :caption: Lib/foo.py
+
+   try:
+       # use the C implementation if possible
+       from fastfoo import greet
+   except ImportError:
+       # fallback to the pure Python implementation
+       def greet():
+           return "Hello World!"
+
+In our example, we need to determine:
+
+- where to place the extension module source code in the CPython project tree;
+- which files to modify in order to compile the CPython project;
+- which ``Makefile`` rules to invoke at the end.
+
+Updating the CPython project tree
+---------------------------------
+
+Usually, accelerator modules are added in the :cpy-file:`Modules` directory of
+the CPython project. If more than one file is needed for the extension module,
+it is more convenient to create a sub-directory in :cpy-file:`Modules`.
+
+For our extension module ``fastfoo``, we consider the following working tree:
+
+- :ref:`Modules/_foo/_foomodule.h` --- the extension module shared prototypes.
```

Comment:
I see. Because most of the modules that are not just a single .c files have a corresponding .h which has a bit of everything inside (almost all other modules have a single HUGE .c which makes it very hard to read in general). 

I'll change this one then!